diff -uprN linux-4.4.194/drivers/media/i2c/nuvoton/nuc980_cap0.c NUC980-linux-4.4.y-master/drivers/media/i2c/nuvoton/nuc980_cap0.c
--- linux-4.4.194/drivers/media/i2c/nuvoton/nuc980_cap0.c	2021-08-05 23:48:46.559220000 -0700
+++ NUC980-linux-4.4.y-master/drivers/media/i2c/nuvoton/nuc980_cap0.c	2021-06-20 19:14:48.000000000 -0700
@@ -1431,7 +1431,7 @@ static irqreturn_t nuvoton_vdi0_isr(int
 	}
 
 	spin_lock(&cam->queue_lock);
-	if((*f)->state == F_QUEUED) {
+	if((*f)->state == F_QUEUED && !list_is_last(&(*f)->frame,&cam->inqueue) ) {
 		list_move_tail(&(*f)->frame, &cam->outqueue);
 	}
 	if (!list_empty(&cam->inqueue)) {
diff -uprN linux-4.4.194/drivers/media/i2c/nuvoton/nuc980_cap1.c NUC980-linux-4.4.y-master/drivers/media/i2c/nuvoton/nuc980_cap1.c
--- linux-4.4.194/drivers/media/i2c/nuvoton/nuc980_cap1.c	2021-08-05 23:48:46.559220000 -0700
+++ NUC980-linux-4.4.y-master/drivers/media/i2c/nuvoton/nuc980_cap1.c	2021-06-20 19:14:48.000000000 -0700
@@ -1416,7 +1416,7 @@ static irqreturn_t nuvoton_vdi1_isr(int
 		}
 	}
 	spin_lock(&cam->queue_lock );
-	if((*f)->state == F_QUEUED) {
+	if((*f)->state == F_QUEUED && !list_is_last(&(*f)->frame,&cam->inqueue)) {
 		list_move_tail(&(*f)->frame, &cam->outqueue);
 	}
 	if (!list_empty(&cam->inqueue)) {
